crates/bifrost/src/providers/replicated_loglet/tasks/find_tail.rs

```diff
@@ -532,7 +546,94 @@ impl<T: TransportConnect> FindTailTask<T> {
+                        //
+                        // sequencer is dead AND:
+                        // - We have f-majority responses already
+                        // - We have write-quorum responses (i.e. tail is repairable)
+                        //
+                        // This means that waiting for nodes to move their local tail is:
+                        // A) unlikely to happen (because sequencer is observed dead)
+                        // B) and if we sealed, we'll be able to repair this tail since we can
+                        // reach to write-quorum.
+                        //
+                        // Our goal is to detect when we should just go ahead and seal instead of
+                        // waiting for the tail to move.
+                        //
+                        // This **must** take into account the sequencer status. If the sequencer
+                        // node is believed to be alive, we should not step over it. Instead, we continue
+                        // waiting because it could be struggling to replicate the tail still. However,
+                        // if the sequencer is believed to dead, we should seal and repair the tail and
+                        // restart the find-tail procedure if it has been over 5 seconds since the
+                        // last update from any log-server (quiescent state).
+                        //
+                        // This is an attempt to reduce over-reacting solely based on observing a "dead"
+                        // sequencer. A dead sequencer can be observed in benign situations, such
+                        // as:
+                        // 1. On startup, we might not have a good view of the cluster state. We don't
+                        //    want to prematurely preempt the sequencer unless we are confident that
+                        //    it's gone for good. Gone is a state defined by observing a higher
+                        //    generation of the same plain node id. If the sequencer is _gone_, it's
+                        //    okay to be eager to seal.
+                        // 2. We don't attempt to seal unless we have heard from the union of the write-quorum
+                        //    and the f-majority, this protects us against being partitioned with the
+                        //    minority of log-servers in the nodeset.
+                        // 3. If the sequencer is not gone (but dead), we react after being quiescent for 5 seconds.
+                        if (is_sequencer_gone(
+                            &cs,
+                            self.my_params.sequencer,
+                            nodes_config.live_load(),
+                        ) || is_sequencer_likely_gone(
+                            &cs,
+                            self.my_params.sequencer,
+                            last_updated,
+                        )) && nodeset_checker.check_write_quorum(NodeTailStatus::is_known)
```

Comment:
Are we missing `&& nodeset_checker.check_fmajority(NodeTailStatus::is_known).passed()` to make sure that we have heard from f-majority of nodes before we try to seal? The comment reads as if this is one of the conditions. If this is true, then we probably also need to update line 509 with the same condition.

Reply:
it's the invariant held at the beginning of `check_nodeset` loop.